### PR TITLE
Add sparse prune info

### PR DESCRIPTION
**Pull request by kewillford**
This changes adds more information to the user's console output to see what is happening when prune is running.

The one thing I don't really like is that there will be an entry in the sparse log file
```
DehydrateVerb {"Output":null,"ReturnCode":0}
```
since there isn't a `StringBuilder commandOutput` that is capturing the output to write to the logs.
There is a entire log file for the dehydrate that has all the log information though so maybe this is okay.  I could also remove the `"Output":null,` and just have the `ReturnCode` in this case or maybe put the information about the verb that was ran and to look in the logs for that verb.

Updated to have this in the sparse log:
```
DehydrateVerb {"Output":"Check dehydrate logs for output","ReturnCode":0}
```
Also thought about having all the information in the sparse log but that would require more extensive changes by having the base class create the tracer for the verb and all the derived class use that tracer.  Seems that change would be outside the scope of these changes though.

### CODE REVIEW

File: GVFS/GVFS/CommandLine/SparseVerb.cs [hunk 1]
```diff
@@ -240,23 +240,32 @@ protected override void Execute(GVFSEnlistment enlistment)
...
-                            tracer,
-                            verb =>
-                            {
-                                verb.Confirmed = true;
-                                verb.Folders = string.Join(FolderListSeparator, directoriesToDehydrate);
-                            });
-
-                        return verbReturnCode == ReturnCode.Success;
+                        verb.Confirmed = true;
+                        verb.Folders = string.Join(FolderListSeparator, directoriesToDehydrate);
```
Comment (wilbaker) [inline GVFS/GVFS/CommandLine/SparseVerb.cs]: Alternative to consider for passing in `this.Output`.

You could just set it in the verb configuration action here:

```
verb.Output = this.Output
```

And then `ExecuteGVFSVerb` could be updated to only use the newly created writer if one wasn't set by the configure action:

```
    ReturnCode returnCode;
    returnCode = this.Execute<TVerb>(
    this.EnlistmentRootPathParameter,
    verb =>
    {
        configureVerb?.Invoke(verb);
        verb.Output = verb.Output ?? writer;
    });
```

(It would mean we'd occasionally create a `StringWriter` we don't need, but we've always been creating one so it's not making things any worse)

File: GVFS/GVFS/CommandLine/SparseVerb.cs [hunk 2]
```diff
@@ -240,23 +240,32 @@ protected override void Execute(GVFSEnlistment enlistment)
...
+                this.ReportErrorAndExit(tracer, "Failed to prune.");
+            }
+
+            this.WriteMessage(tracer, $"Found {directoriesToDehydrate.Length} folders to prune.");
+
             if (directoriesToDehydrate.Length > 0)
             {
-                if (!this.ShowStatusWhileRunning(
-                    () =>
+                ReturnCode verbReturnCode = this.ExecuteGVFSVerb<DehydrateVerb>(
```
Comment (wilbaker) [inline GVFS/GVFS/CommandLine/SparseVerb.cs]: Should the console messages in Dehydrate either be parameterized or made more general?  

It looks like this message is always printed:

https://github.com/microsoft/VFSForGit/blob/a537d002be8d48675f48e3e97fe1c8c6359133a9/GVFS/GVFS/CommandLine/DehydrateVerb.cs#L166

And there are some other error messages that use the work "dehydrate":

https://github.com/microsoft/VFSForGit/blob/a537d002be8d48675f48e3e97fe1c8c6359133a9/GVFS/GVFS/CommandLine/DehydrateVerb.cs#L209

Should they say prune (or some verb-neutral word)?

File: GVFS/GVFS/CommandLine/SparseVerb.cs (hunk 1 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/SparseVerb.cs]: The `GVFSVerb.Output` gets set in the constructor so I don't think it will every be null unless we make more changes and not set it in the constructor but I would need to see if that would have any bad side effects.
```
        public GVFSVerb(bool validateOrigin = true)
        {
            this.Output = Console.Out;
```

Comment (wilbaker) [reply GVFS/GVFS/CommandLine/SparseVerb.cs]: I totally missed that, thanks!

File: GVFS/GVFS/CommandLine/SparseVerb.cs (hunk 2 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/SparseVerb.cs]: Yeah that would make it easier for the user to understand and we would need possibly different terms for each because you would probably want
`WARNING: If you abort the dehydrate after this point, the repo may become corrupt`
to be
`WARNING: If you abort the sparse after this point, the repo may become corrupt`
but 
`Cannot dehydrate: must have a clean git status.`
to be
`Cannot prune: must have a clean git status.`

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/DehydrateTests.cs [hunk 3]
```diff
@@ -183,7 +183,7 @@ public void FolderDehydrateFolderThatWasEnumerated()
             string subFolderToEnumerate = Path.Combine(pathToEnumerate, "GVFS");
             this.fileSystem.EnumerateDirectory(subFolderToEnumerate);
 
-            this.DehydrateShouldSucceed(new[] { "GVFS folder successfully dehydrated." }, confirm: true, noStatus: false, foldersToDehydrate: "GVFS");
+            this.DehydrateShouldSucceed(new[] { "GVFS folder dehydrate successful." }, confirm: true, noStatus: false, foldersToDehydrate: "GVFS");
```
Comment (derrickstolee) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/DehydrateTests.cs]: Should this string be a `static readonly` in the class?

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/DehydrateTests.cs [hunk 4]
```diff
@@ -329,7 +329,7 @@ public void FolderDehydrateNestedFoldersChildBeforeParent()
             this.fileSystem.ReadAllText(fileToRead2);
 
             this.DehydrateShouldSucceed(
-                new[] { $"{folderToDehydrate1} folder successfully dehydrated.", $"{folderToDehydrate2} folder successfully dehydrated." },
+                new[] { $"{folderToDehydrate1} folder dehydrate successful.", $"{folderToDehydrate2} folder dehydrate successful." },
```
Comment (derrickstolee) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/DehydrateTests.cs]: Should these strings also be a constant? 

```
static readonly string FolderDehydrateSuccessfulFormat = "{0} folder dehydrate successful.";

...

this.DehydrateShouldSucceed(
    new []
    {
        String.Format(FolderDehydrateSuccessfulFormat, folderToDehydrate1), 
        String.Format(FolderDehydrateSuccessfulFormat, folderToDehydrate2),
    },
    confirm: true,
    noStatus: false,
    foldersToDehydrate: string.Join(";", folderToDehydrate1, foldertoDehydrate2));
```

**Comment by kewillford** [comment]
azp run PR - Windows - Functional Tests (Sparse Mode)

**Comment by kewillford** [comment]
azp run PR - Windows - Functional Tests (Sparse Mode)

File: GVFS/GVFS/CommandLine/DehydrateVerb.cs [hunk 5]
```diff
@@ -334,12 +349,13 @@ private void SendDehydrateMessage(ITracer tracer, GVFSEnlistment enlistment, str
...
                 {
-                    this.WriteMessage(tracer, $"{folder} folder successfully dehydrated.");
+                    this.WriteMessage(tracer, $"{folder} folder {this.DisplayName} successful.");
                 }
 
                 foreach (string folder in response.FailedFolders)
                 {
-                    this.WriteMessage(tracer, $"{folder} folder failed to dehydrate. You may need to reset the working directory by deleting {folder}, running `git reset --hard`, and retry the dehydrate.");
+                    this.WriteMessage(tracer, $"{folder} folder failed to {this.DisplayName}. You may need to reset the working directory by deleting {folder}, running `git reset --hard`, and retry the {this.DisplayName}.");
+                    folderErrors.Add(folder);
```
Comment (wilbaker) [inline GVFS/GVFS/CommandLine/DehydrateVerb.cs]: Is `\0{ioError}` included in the entries in `response.FailedFolders`?

If not, should we an a `\0<error>` here?

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 6]
```diff
@@ -887,19 +889,23 @@ protected ReturnCode ExecuteGVFSVerb<TVerb>(ITracer tracer, Action<TVerb> config
...
-                        { "ReturnCode", returnCode }
-                        });
+                    EventMetadata metadata = new EventMetadata();
+                    if (outputWriter == null)
+                    {
+                        metadata.Add("Output", commandOutput.ToString());
+                    }
+                    else
+                    {
+                        metadata.Add("Output", $"Check {new TVerb().VerbName} logs for output");
```
Comment (wilbaker) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: I don't quite follow this check.  

If `outputWriter` is non-null (e.g. it has been set to `Console.Out`), how is that related to the verb log containing the output?

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 7]
```diff
@@ -92,6 +93,7 @@ public string ServicePipeName
         }
 
         public TextWriter Output { get; set; }
+        public TextWriter ErrorOutput { get; set; }
```
Comment (wilbaker) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: Should this only be used for machine readable output?  Or in the future should new error messages be written to `ErrorOutput`?

File: GVFS/GVFS/CommandLine/DehydrateVerb.cs [hunk 8]
```diff
@@ -260,9 +264,10 @@ private void DehydrateFolders(JsonTracer tracer, GVFSEnlistment enlistment, stri
                                     {
                                         if (!this.TryIO(tracer, () => this.fileSystem.DeleteDirectory(fullPath), $"Deleting '{fullPath}'", out ioError))
                                         {
-                                            this.WriteMessage(tracer, $"Cannot dehydrate folder '{folder}': removing '{folder}' failed.");
+                                            this.WriteMessage(tracer, $"Cannot {this.DisplayName} folder '{folder}': removing '{folder}' failed.");
                                             this.WriteMessage(tracer, "Ensure no applications are accessing the folder and retry.");
                                             this.WriteMessage(tracer, $"More details: {ioError}");
+                                            folderErrors.Add($"{folder}\0{ioError}");
```
[the remark this answers is not in the document]
Comment (kewillford) [reply GVFS/GVFS/CommandLine/DehydrateVerb.cs]: Looks like a space
```
Scripts The process cannot access the file 'C:\gvfs_repos\ft\src\Scripts' because it is being used by another process.
Failed to dehydrate 1 folder(s).
Failed to prune. Exit Code: DehydrateFolderFailures
```

File: GVFS/GVFS/CommandLine/DehydrateVerb.cs (hunk 5 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/DehydrateVerb.cs]: These folders come back from the mount process and I wasn't sure how a `\0` would be handled by our named pipe messages but let me take a look and try it out.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 7 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: I really don't think we want to put that restriction on this.  We haven't been writing to stderr and it isn't displayed by default in the console.  If we want to really create a machine readable CLI then we should do something else but not sure how much work we want to put into it.

Comment (wilbaker) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: > it isn't displayed by default in the console

I did not realize this thanks!

Sounds like this change is good as it is now.

File: GVFS/GVFS/CommandLine/DehydrateVerb.cs (hunk 5 above)
Comment (wilbaker) [reply GVFS/GVFS/CommandLine/DehydrateVerb.cs]: Basic question, is the `\0<error>` just to provide more details to the caller (in case they care about the specific error reason)?

I wasn't sure if callers would expect every line to have a `\0<error>` even if it was just ` `\0{DisplayName} Failed` (or some other general failure message for this case)

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 6 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: The case I am trying to handle is the `sparse` command running the `DehyrateVerb` which the output is being redirected to `Console.Out` but it creates it's own log file for logging information so if you only look in the `SparseVerb` log you may not see everything that happened during the dehydrate.  That is why it is adding this to the sparse log so you know to look at the log file of the `new TVerb().VerbName` for details.

sparse log will have something like
```
sparse {"Message":"No folders to update in sparse set."}
sparse {"Message":"Found 1 folders to prune."}
DehydrateVerb {"Output":"Check dehydrate logs for output","ReturnCode":7}
Error {"ErrorMessage":"Failed to prune. Exit Code: DehydrateFolderFailures"}
```

Dehydrate log will have the full logs from the verb including
```
Error {"Description":"Deleting '<folder path>'","Error":"The process cannot access the file '<folder path>' because it is being used by another process.","ErrorMessage":"TryIO: Caught exception performing action"}
```

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 7 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: I will still add a comment about how we are using this right now in case something else wants to use it.

Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: Sorry it's not Console doesn't show stderr.  It does but I wasn't seeing the stderr stream from the DehydrateVerb in the sparse console output without using `2>&1`, so maybe there is something more I need to do here.

Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: Weird it is working the way you would expect.  Not sure what was going on when I didn't see the folder list in the console, maybe I didn't have everything connected yet.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 6 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: Making the entire `TVerb` output logs go to the calling verb's log file is more work since the `ITracer` would need to be passed around and that seemed out of scope for these changes.

File: GVFS/GVFS/CommandLine/DehydrateVerb.cs (hunk 5 above)
Comment (kewillford) [reply GVFS/GVFS/CommandLine/DehydrateVerb.cs]: Yeah we should always have at least the `\0` even if there isn't anything after so that it is an expected format where the `\0` is the field separator.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 6 above)
Comment (wilbaker) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: > but it creates it's own log file for logging information

To confirm, in this example will `sparse`'s log file include the contents of whatever was written to `outputWriter`?  Or is this never recorded by a tracer?

This might be a dumb question but I can't figure out where this:

`{ "Output", commandOutput.ToString() },`

Was ever getting logged (e.g. I don't see it in the `clone` verb when it calls `mount`).

> you may not see everything

Maybe add a comment here, something like:

```
else
{
    // If a parent verb is redirecting the output of its child, include a reminder
    // that the child verb's activity was recorded in its own log file
    metadata.Add("Output", $"Check {new TVerb().VerbName} logs for output");
```

Comment (kewillford) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: >To confirm, in this example will sparse's log file include the contents of whatever was written to outputWriter? Or is this never recorded by a tracer?

No, writing to the `outputWriter` doesn't have anything to do with the `ITracer` that is being used for tracing.  That is what makes this kind of messy because we are redirecting where output is going (to the `Console.Out` of the sparse verb) but the `DehyrateVerb` is still creating its own tracer that is used and writes to the dehydrate log.

>Was ever getting logged (e.g. I don't see it in the clone verb when it calls mount).

Looks like in the case of the `CloneVerb` its tracer is being disposed before the call to `this.Execute<MountVerb>(`.
See https://github.com/microsoft/VFSForGit/blob/c5b4262e38870eb4c648a9bac48ca3c062aad5bc/GVFS/GVFS/CommandLine/CloneVerb.cs#L123-L204

Comment (wilbaker) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: > Looks like in the case of the CloneVerb its tracer is being disposed before the call to this.Execute<MountVerb>

Ahh that does sound familiar there was one point in time where we could not have two ETW tracer instances at the same time.  I don't think we still have that restriction but this code was never updated.

> That is what makes this kind of messy because we are redirecting where output is going (to the Console.Out of the sparse verb) but the DehyrateVerb is still creating its own tracer that is used and writes to the dehydrate log.

Thanks for all of the details here!  The `{ "Output", commandOutput.ToString() }` was what was really throwing me off, but it looks like it's probably never gone where we expected (after updating `clone` to dispose its tracer).

The latest changes look good to me, thanks!
